lib/usagereporter/teleport/aggregating/reporter.go

```diff
@@ -0,0 +1,279 @@
+	"github.com/gravitational/teleport/lib/backend"
+	usagereporter "github.com/gravitational/teleport/lib/usagereporter/teleport"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+const (
+	reportGranularity = 15 * time.Minute
+	rollbackGrace     = time.Minute
+	reportTTL         = 60 * 24 * time.Hour
+)
+
+type ReporterConfig struct {
+	Backend backend.Backend
+	Log     logrus.FieldLogger
+
+	ClusterName    types.ClusterName
+	ReporterHostID string
+}
+
+func (cfg *ReporterConfig) CheckAndSetDefaults() error {
+	if cfg.Backend == nil {
+		return trace.BadParameter("missing Backend")
+	}
+	if cfg.Log == nil {
+		return trace.BadParameter("missing Log")
+	}
+	if cfg.ClusterName == nil {
+		return trace.BadParameter("missing ClusterName")
+	}
+	if cfg.ReporterHostID == "" {
+		return trace.BadParameter("missing ReporterHostID")
+	}
+	return nil
+}
+
+// NewReporter returns a new running [Reporter]. To avoid resource leaks,
+// GracefulStop must be called or the base context must be closed. The context
+// will be used for all backend operations.
+func NewReporter(ctx context.Context, cfg ReporterConfig) (*Reporter, error) {
+	if err := cfg.CheckAndSetDefaults(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	anonymizer, err := utils.NewHMACAnonymizer(cfg.ClusterName.GetClusterID())
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	baseCtx, baseCancel := context.WithCancel(ctx)
+
+	r := &Reporter{
+		anonymizer: anonymizer,
+		svc:        reportService{cfg.Backend},
+		log:        cfg.Log,
+
+		ingest:  make(chan usagereporter.Anonymizable),
+		closing: make(chan struct{}),
+		done:    make(chan struct{}),
+
+		clusterName:    anonymizer.AnonymizeNonEmpty(cfg.ClusterName.GetClusterName()),
+		reporterHostID: anonymizer.AnonymizeNonEmpty(cfg.ReporterHostID),
+
+		baseCancel: baseCancel,
+	}
+
+	go r.run(baseCtx)
+
+	return r, nil
+}
+
+// Reporter aggregates and persists usage events to the backend.
+type Reporter struct {
+	anonymizer utils.Anonymizer
+	svc        reportService
+	log        logrus.FieldLogger
+
+	// ingest collects events from calls to [AnonymizeAndSubmit] to the
+	// background goroutine.
+	ingest chan usagereporter.Anonymizable
+	// closing is closed when we're not interested in collecting events anymore.
+	closing chan struct{}
+	// closingOnce closes [closing] once.
+	closingOnce sync.Once
+	// done is closed at the end of the background goroutine.
+	done chan struct{}
+
+	// clusterName is the anonymized cluster name.
+	clusterName []byte
+	// reporterHostID is the anonymized host ID of the reporter (this agent).
+	reporterHostID []byte
+
+	// baseCancel cancels the context used by the background goroutine.
+	baseCancel context.CancelFunc
+}
+
+var _ usagereporter.GracefulStopper = (*Reporter)(nil)
+
+// GracefulStop implements [usagereporter.GracefulStopper].
+func (r *Reporter) GracefulStop(ctx context.Context) error {
+	r.closingOnce.Do(func() { close(r.closing) })
+
+	select {
```

Comment:
I don't understand the semantics here - what is the passed in context used for?

Reply:
Documented in the interface in 91b75dc06280c0406ce34c81da44722698b5c93c:
```go
	// GracefulStop gracefully closes and runs any finalization needed by the
	// UsageReporter; operations can run as long as the context is alive, but
	// must terminate quickly (even losing data) if the context is closed.
	// Returns nil if operations have completed cleanly.
	GracefulStop(context.Context) error
```

Reply:
In the case of the `aggregating.Reporter`, closing the context that's passed in will trigger the closing of the base context, and thus all closing operations ungracefully (and quickly) fail.